### ISSUE TITLE
Questions about HW-aware training and inference

### ISSUE THREAD

**Issue by darrenkao** [IBM/aihwkit#275]
Are inp_res and out_res don't affect accuracy in inference stage?
I use an `InferenceRPUConfig` with `forward.is_perfect = True` to train a ideal model. After that, I load the model to with `forward.inp_res = 1/32`, `forward.out_res = 1/32`, but accuracy didn't drop. The same situation even as I set them to 1. I observe that in inference stage, the noise only controlled  by `model.drift_analog_weight()`, right?

**Comment by maljoras** [comment]
Hi @darrenkao,
Thanks again for the question. Actually, you raise a good point, in fact, we store the rpu config when saving the state dict, and when one loads again, the current rpu config of the model is overwritten with the stored one. So what you are seeing is the effect of the rpu config still being perfect, because you loaded the state dict from the stored perfect model.

For inference and hardware aware training it would indeed be useful to train and infer with different RPU configs, which is currently not possible, because the stored rpu config is applied automatically. I will add a switch in the load statedict method, so that one can turn off the auto applying of the rpu config at least for inference rpu config where it makes sense.

Thanks for raising the issue!

**Comment by maljoras** [comment]
Hi @darrenkao,
thanks again for raising this issue. I have now added a switch to be able to change the RPU Config, in the way you were doing the experiments. 

**Example:**
 
Assume, `create_model` creates a model using a certain `rpu_config`. One can now eg:
```python
model = create_model(rpu_config=SingleRPUConfig(
                                  forward=IOParameters(is_perfect=True),     
                                  backward=IOParameters(is_perfect=True)))
                                    
# [... train model ... ]       
state_dict = model.state_dict()
      
# change rpu_config to non-perfect
new_model= create_model(rpu_config=SingleRPUConfig())
        
# apply weights, etc. but change the backward/forward HW configuration 
new_model.load_state_dict(state_dict, load_rpu_config=False)

# [... evaluate new model ... ]
```       
In this case, the new model would have the RPU Config with `is_perfect=False` which would not be overwritten by the stored RPU Config in the `state_dict`. Note that the latter is the default behavior and the reason why you saw the issue in the first place.
